GitCommands/CommitMessageManager.cs

```diff
@@ -0,0 +1,124 @@
+﻿using System.IO.Abstractions;
+using System.Text;
+
+namespace GitCommands
+{
+    public interface ICommitMessageManager
+    {
+        /// <summary>
+        /// Reads/stores whether the previous commit shall be amended (if AppSettings.RememberAmendCommitState).
+        /// </summary>
+        bool AmendState { get; set; }
+
+        /// <summary>
+        /// The pathname of the file where a prepared (non-merge) commit message is stored.
+        /// </summary>
+        string CommitMessagePath { get; }
+
+        /// <summary>
+        /// Returns whether .git/MERGE_MSG exists.
+        /// </summary>
+        bool IsMergeCommit { get; }
+
+        /// <summary>
+        /// Reads/stores the prepared commit message from/in .git/MERGE_MSG if it exists or else in .git/COMMITMESSAGE.
+        /// </summary>
+        string MergeOrCommitMessage { get; set; }
+
+        /// <summary>
+        /// Deletes .git/COMMITMESSAGE and the file with the AmendState.
+        /// </summary>
+        void ResetCommitMessage();
+    }
+
+    public sealed class CommitMessageManager : ICommitMessageManager
+    {
+        private readonly string _amendSaveStatePath;
+        private readonly string _commitMessagePath;
+        private readonly string _mergeMessagePath;
+
+        private Encoding _commitEncoding;
+        private IFileSystem _fileSystem;
+
+        public CommitMessageManager(string workingDirGitDir, Encoding commitEncoding)
+            : this(workingDirGitDir, commitEncoding, new FileSystem())
+        {
+        }
+
+        private CommitMessageManager(string workingDirGitDir, Encoding commitEncoding, IFileSystem fileSystem)
+        {
+            _fileSystem = fileSystem;
+            _commitEncoding = commitEncoding;
+            _amendSaveStatePath = GetFilePath(workingDirGitDir, "GitExtensions.amend");
+            _commitMessagePath = GetFilePath(workingDirGitDir, "COMMITMESSAGE");
+            _mergeMessagePath = GetFilePath(workingDirGitDir, "MERGE_MSG");
```

Comment:
🤔  Perhaps it is worth to use an instance of `IFullPathResolver` here constructed as
```
var resolver = new FullPathResolver(() => workingDirGitDir)
```

Reply:
I understand your point.
Though that's where you guided me to: https://github.com/gitextensions/gitextensions/pull/6530#pullrequestreview-233799973. A lambda `Func<string> getWorkingDirGitDir` would have the same problem and is less visible IMO.
And I think, a change of the repo needs consideration everywhere in GE at all. The lifetime of the instance is bound to FormCommit and it's not very likely to be used in a different manner.

If you insist, I would always create new instances of the `CommitMessageManager` and hide this behind a static interface (facade).

> Perhaps it is worth to use an instance of IFullPathResolver.

(There are two known bugs, which shouldn't have a negative effect on accessing files in the `.git` folder.)
My main concern is that the lambda will be hidden even more.

Reply:
This isn't a blocking comment, more like a discussion point. I'm always looking at how we can reduce a duplication of the code, especially when we have existing implementations already tested.
Perhaps I worry too much :) and you're right, the _current_ context is scoped to `FormCommit`, so your implementation is fine.

My original suggestion was about not passing an instance of `IGitModule`, as it may lead to operating on a different repo to what a user may be currently in.